mlcomp/parallelm/ml_engine/sagemaker_engine.py

```diff
@@ -0,0 +1,111 @@
+
+from parallelm.common import constants
+from parallelm.common.mlcomp_exception import MLCompException
+from parallelm.ml_engine.ee_arg import EeArg
+from parallelm.ml_engine.python_engine import PythonEngine
+
+
+class SageMakerEngine(PythonEngine):
+    def __init__(self, pipeline):
+        super(SageMakerEngine, self).__init__(pipeline, None)
+        self._iam_role = None
+        self._iam_role_name = None
+
+        eng_args_config = self._read_execution_env_params()
+
+        self._tag_key = EeArg(eng_args_config.get('tag-key')).value
+        self._tag_value = EeArg(eng_args_config.get('tag-value')).value
+
+        self._setup_env(eng_args_config)
+
+        # Note: the 'boto3' module should be imported only after we setup the environment variables,
+        # which setup the shared credentials and region configuration
+        global boto3, ClientError
+        import boto3
+        from botocore.exceptions import ClientError
+
+        self._setup_logger(eng_args_config)
+        self._setup_iam_role(eng_args_config)
+
+    @property
+    def iam_role(self):
+        return self._iam_role
+
+    def _setup_iam_role(self, eng_args_config):
+        self._iam_role = EeArg(eng_args_config.get('iam_role')).value
+        if not self._iam_role:
+            # acount_id = boto3.client('sts').get_caller_identity().get('Account')
+
+            now = datetime.datetime.now()
+            role_name = 'AmazonSageMaker-ExecutionRole-{y}{month}{d}T{h}{minute}{s}' \
+                .format(y=now.year, month=now.month, d=now.day, h=now.hour, minute=now.minute, s=now.second)
+
+            tags = [{'Key': self._tag_key, 'Value': self._tag_value}] if self._tag_key else []
+
+            client = boto3.client('iam')
+
+            try:
+                response = client.create_role(
+                    Path='/service-role/',
+                    RoleName=role_name,
+                    AssumeRolePolicyDocument='<URL-encoded-JSON>',
+                    Description='Auto generated sagemaker aim role by ParallelM',
+                    Tags=tags
+                )
+                self._iam_role_name = response['Role']['RoleName']
+                self._iam_role = response['Role']['Arn']
+            except ClientError as e:
+                self._logger.error("Failed to create a an iam role for sagemaker service!\n{}".format(e))
+                raise e
+
+    def _read_execution_env_params(self):
+        ee_config = self._pipeline.get('executionEnvironment', dict()).get('configs')
+        if not ee_config:
+            raise MLCompException("Missing execution environment section in pipeline json!")
+
+        eng_config = ee_config.get('engConfig')
+        if not eng_config:
+            raise MLCompException("Missing execution environment engine section in pipeline json!")
+
+        if eng_config['type'] != 'sagemaker':
+            raise MLCompException("Unexpected engine type in execution environment! expected: 'sagemake', got: {}"
```

Comment:
Maybe we should introduce MLCompSageMakerEngineException?

Reply:
What would be the benefit of it? Do we really need to distinguish such exception from others?

Reply:
General rule is to make exception as specific is possible, so I would create specific exception for every engine.

Reply:
Ok. In such case, we'll better organize the exceptions files hierarchy and instead of putting them all directly under `common`, we rather add `exceptions` dir under common and put all exceptions files there.
If I'll do it now, it'll be involved is massive refactoring in `mlcomp` and in `mlhub`. I suggest to leave it to another task.

Reply:
Created new task: https://parallelmachines.atlassian.net/browse/REF-6184